### PR TITLE
[WIP] Optimizer design alternative

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Describe your changes.

After internal discussion, we split this optimizer design work into few stages, will create separated PRs one by one. 

Close this draft PR. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
More information required or a reference to a design document for fundamental changes like this. Having a vector of other OrtValues inside the OrtValues introduces recursiion, memory allocation. The vector is not even optional.

**Comment by pengwa** [comment]
> More information required or a reference to a design document for fundamental changes like this. Having a vector of other OrtValues inside the OrtValues introduces recursiion, memory allocation. The vector is not even optional. The feature of OrtValue being copied cheaply and still referring to the same buffer is no longer held. This applies recursively.

 Thanks for the attention. Yeah, you are right. I was going to take this into an internal discussion. The implementation detail we took here is not elegant I admit, which is something we need improve, if once we have a conclusion doing such a InplaceToSequence thing after the discussion.

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -1047,6 +1047,155 @@ void RegisterTrainingOpSchemas() {
...
+          "Constrain learning rate to float")
+      .TypeConstraint(
+          "T2",
+          {"int64"},
+          "Constrain step count to 64-bit integer");
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(InPlaceToSequence)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc("")
```
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1139]: Add description for the op.

Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1139]: Add description for the op.

Comment (askhade) [reply orttraining/orttraining/core/graph/training_op_defs.cc:1139]: How is this different than SequenceConstruct? https://github.com/onnx/onnx/blob/main/docs/Operators.md#sequenceconstruct

Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:1139]: It is trying the reuse all inputs' memory (OrtValues), the dimplementation of SequenceConstruct in current ORT is doing copy. We need optimizer in-place operater the weight update

**Pull request closed without merging**